### PR TITLE
Weather Man app with proper documentation.

### PR DESCRIPTION
**Pull request by uzair-imtiaz**
### Description:
This application generates reports for the weather status in Murree. The weather data is available in individual files, with each file containing temperature, dew point, humidity, and other relevant values for a specific day of a month.
The script allows you to generate reports by providing command line arguments: the path to the directory where the weather files are located, the report type you want to generate, and the year and month for which you want the report. The generated report will be displayed on the console.

### Testing Instructions:
To test the updated functionality, please follow the instructions below:
1. Create a virtual environment:
`python -m venv env`

2. Activate the virtual environment:

   - For Windows:
      `.\env\Scripts\activate`

   - For macOS/Linux:
      `source ./env/bin/activate`
3. Run the following command to generate a report:
`main.py /path/to/files-dir -a 2009/1`

**Note:** Replace /path/to/files-dir with the actual path to the directory containing the weather files. If the directory is in the same location as main.py, you can use weatherfiles directly.

### Expected output:

> ----Monthly Report----
> 
> Highest Average: 10C
> Lowest Average: 2C
> Average Mean Humidity: 50%

You can request more than one report simultaneously.
`main.py /path/to/files-dir -c 2011/03 -a 2011/3 -e 2011`
Replace /path/to/files-dir with the actual path to the directory containing the weather files.

### Snapshots of sample runs of the program:
Screenshot 2023-06-26 at 5 02 56 PM

Screenshot 2023-06-26 at 5 06 21 PM

Screenshot 2023-06-26 at 5 07 38 PM

### CODE REVIEW

File: ComputeResults.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class ComputeResults:
+    """
+    A class that will compute the results for the reports and store those in the calculationResults type object.
+
+    Methods
+    -------
+    assign_to_results_class
+        Calculates the weather readig values and assign them to a CalculationResult type object.
```
Comment (Ali-D-Akbar) [inline ComputeResults.py]: spell: reading

Comment (Ali-D-Akbar) [reply ComputeResults.py]: spell: assigns them

File: CalculationsResult.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
+
+
+class CalculationsResult:
+    """A class that will hold the calculations results.
+
+    Attributes
+    ----------
+    average_mean_humidity : int >= 0
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: average and mean both are the same words. Can be `average_of_mean_humidity`

File: CalculationsResult.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+
+
+class CalculationsResult:
+    """A class that will hold the calculations results.
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: nit: calculation results
Perhaps the class name can also be updated

File: ComputeResults.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def assign_to_results_class(self, weather_readings) -> CalculationsResult:
```
Comment (Ali-D-Akbar) [inline ComputeResults.py]: or simply write `assign_results`

File: ComputeResults.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        Parameters
+        ----------
+        weather_readings : list
+            A list of weather reading values of a month.
+
+        Returns
+        -------
+        calculations : CalculationsResult
+            A CalculationsResult object with maximum temperature and the date that day, minimum temperature and the date
```
Comment (Ali-D-Akbar) [inline ComputeResults.py]: nit: the date of the day

File: ComputeResults.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            that day, maximum humidity ad the date that day assigned by respective values from that month.
```
Comment (Ali-D-Akbar) [inline ComputeResults.py]: spell: and

File: ComputeResults.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        """
+
+        # Using the lambda function to pass an extra parameter for the max function. That extra parameter is the value
+        # at the index of that specific reading.
+        max_temp_entry = max(weather_readings, key=lambda x: x.temperature.max)
```
Comment (Ali-D-Akbar) [inline ComputeResults.py:50]: nice way to use lambda functions.

File: FileParser.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+        Returns
+        -------
+        weather_readings : list
+            The list of weatherReadings type object populated with readings data from the file.
+        """
+
+        weather_readings = []
+        for reading in readings_list:
+            w_reading = WeatherReadings(reading)
```
Comment (Ali-D-Akbar) [inline FileParser.py]: Use complete variable names. Always. `weather_reading`

File: ReportGenerator.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
+import glob
+from FileParser import FileParser
+from ComputeResults import ComputeResults
+from datetime import datetime
+
+
+class ReportGenerator:
+    """A class that will generate reports.
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: add a linebreak after """

File: CalculationsResult.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: Docstrings missing for all the files.

File: ReportGenerator.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    Methods
+    -------
+    generate_report
+        Calls the appropriate functions according to the CLI arguments.
+
+    report_e
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: use complete names instead of single letters.

File: ReportGenerator.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        For a given year display the highest and lowest temperature and day, most humid day and humidity.
+
+        Parameters
+        ----------
+        files : list
+            A list of files with name similar to the one passed by the user.
+            Must contain at one value.
+        """
+
+        fp = FileParser(files[0])
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: use complete name `file_parser`

File: ReportGenerator.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        readings_list = fp.parse_file()
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: nit: `reading_list`

File: main.py [hunk 14]
```diff
@@ -0,0 +1,17 @@
+from ReportGenerator import ReportGenerator
+from arg_parser import parse_args
+
+
+def main():
+    """The main function.
```
Comment (Ali-D-Akbar) [inline main.py]: linebreak after """

File: CalculationsResult.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    highest_temperature : int >= 0
+         Holds the maximum temperature value of the month.
+
+    most_humid_day : str
+        Date when the most humidity was recorded in string format.
+
+    Methods
+    -------
+    set_extremes
+        Assign the extreme(highest, lowest) values of temperature ad humidity ad record the date when it occurred.
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: is your `n` button missing? 😅

File: CalculationsResult.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    Attributes
+    ----------
+    average_mean_humidity : int >= 0
+        Holds the average of the mean humidity of the month.
+
+    lowest_average_temperature : int >= 0
+        Holds the minimum average from the readings of the month.
+
+    highest_average_temperature : int >= 0
+        Holds the minimum average from the readings of the month.
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: Duplicate messages. Please fix this.

File: CalculationsResult.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
+
+
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: remove extra spaces

File: CalculationsResult.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+            Date when the maximum humidity was recorded in string format.
+        """
+        self.highest_temperature = highest
+        self.highest_temperature_day = day_of_highest
+        self.lowest_temperature = lowest
+        self.lowest_temperature_day = day_of_lowest
+        self.humidity = humidity
+        self.most_humid_day = most_humid_day
+
+    def set_average(self, highest_avg_temperature, lowest_avg_temperature, avg_mean_humidity) -> None:
```
Comment (Ali-D-Akbar) [inline CalculationsResult.py]: nit: set_averages

File: ReportGenerator.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        formatted_date = datetime.strptime(selected_results.lowest_temperature_day,
+                                           "%Y-%m-%d").strftime("%B %d")
+        print('Lowest: {}C on {}'.format(selected_results.lowest_temperature,
+                                         formatted_date))
+        formatted_date = datetime.strptime(selected_results.most_humid_day,
+                                           "%Y-%m-%d").strftime("%B %d")
+        print('Humidity: {}% on {}'.format(selected_results.humidity,
+                                           formatted_date))
+
+    def report_a(self, file) -> None:
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: Perhaps these names name be improved.

File: arg_parser.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    ------
+    ParserError
+        if given arguments are not of the correct format.
+    """
+
+    parser = argparse.ArgumentParser(prog='Weather Man.', description='Generates a report for the specified month and'
+                                                                      'year for Murree.')
+    parser.add_argument('files_dir', help='Path to the files directory')
+    parser.add_argument('-a', '--year_and_month_for_a', help='Report a requires year ad month in yyyy/mm format')
+    parser.add_argument('-c', '--year-and-month-for-c', help='Report c requires year ad month in yyyy/mm format')
```
Comment (Ali-D-Akbar) [inline arg_parser.py]: use hiphens for arguments.

**Review by Ali-D-Akbar — CHANGES_REQUESTED** [review]
Good 1st PR 🎉 
Please add a `.gitignore` file in your repo to avoid pushing files like `.DS_Store`. You can find standard gitignore files all over the internet. 
Also, update the description to have a couple of headings like `Description`, `Testing Instructions` and some screenshots. Think of me as a total newbie and add your testing instructions like this:

1. Create virtual environment: `python -m venv env`
2. Activate virtual environment: `source ./venv/bin/activate`
3. Run the following command: `main.py loremipsum`
4. Expect an output like `yada yada yada`

File: .gitignore [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+# Ignore the `build` directory if applicable
+/build/
+
+# Ignore any sensitive or environment-specific files
+secret.txt
+config.ini
+credentials.json
+
+# Ignore Jupyter Notebook checkpoints
+.ipynb_checkpoints/
```
Comment (Ali-D-Akbar) [inline .gitignore]: Why is this file pasted twice?

File: ReportGenerator.py (hunk 11 above)
Comment (Ali-D-Akbar) [reply ReportGenerator.py]: @Uzair-Imtiaz this hasn't been addressed yet.

File: ReportStats.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+    highest_temperature : int >= 0
+         Holds the maximum temperature value of the month.
+
+    most_humid_day : str
+        Date when the most humidity was recorded in string format.
+
+    Methods
+    -------
+    set_extremes
+        Assign the extreme(highest, lowest) values of temperature ad humidity ad record the date when it occurred.
```
Comment (Ali-D-Akbar) [inline ReportStats.py]: Still not addressed yet. Please look at your code to find all the typos.

File: ReportStats.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+        self.humidity = None
+        self.lowest_temperature_day = None
+        self.lowest_temperature = None
+        self.highest_temperature_day = None
+        self.highest_temperature = None
+        self.most_humid_day = None
+
+    def set_extremes(self, highest, day_of_highest, lowest, day_of_lowest, humidity, most_humid_day) -> None:
+        """
+        Assign the extreme(highest, lowest) values of temperature ad humidity ad record the date when it occurred.
```
Comment (Ali-D-Akbar) [inline ReportStats.py]: typo

File: arg_parser.py [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+    ------
+    ParserError
+        if given arguments are not of the correct format.
+    """
+
+    parser = argparse.ArgumentParser(prog='Weather Man.', description='Generates a report for the specified month and'
+                                                                      'year for Murree.')
+    parser.add_argument('files_dir', help='Path to the files directory')
+    parser.add_argument('-a', '--year-and-month-for-a', help='Report a requires year ad month in yyyy/mm format')
+    parser.add_argument('-c', '--year-and-month-for-c', help='Report c requires year ad month in yyyy/mm format')
```
Comment (Ali-D-Akbar) [inline arg_parser.py]: typos

File: ComputeResults.py [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+            sum_highest_temp += reading.temperature.max
+            sum_lowest_temp += reading.temperature.min
+            sum_mean_humidity += reading.humidity.mean
+        avg_highest_temp = sum_highest_temp // len(weather_readings)
+        avg_lowest_temp = sum_lowest_temp // len(weather_readings)
+        avg_mean_humidity = sum_mean_humidity // len(weather_readings)
+        calculations = ReportStats()
+        calculations.set_extremes(max_temp, max_temp_date, min_temp, min_temp_date, max_humidity, max_humidity_date)
+        calculations.set_averages(avg_highest_temp, avg_lowest_temp, avg_mean_humidity)
+        return calculations
```
Comment (Ali-D-Akbar) [inline ComputeResults.py]: Perhaps the readability can be improved by grouping a bunch of statements together

File: ReportGenerator.py [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+Usage
+-----
+    from ReportGenerator import ReportGenerator
+
+    # Create an instance of ReportGenerator
+    report_generator = ReportGenerator()
+
+    # Generate the report based on the provided arguments
+    report_generator.generate_report(*args)
+"""
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py]: You can remove the filename from docstring. Also for the next time, no need to add a lot of info. One or two lines describing a file is sufficient.

**Review by Ali-D-Akbar — COMMENTED** [review]
Some readability can be improved by linebreaking and grouping a couple of statements. Just a bunch of more changes before approval.

File: .gitignore (hunk 21 above)
Comment (uzair-imtiaz) [reply .gitignore]: The keyboard needs to be taken care of. I'll have the IT team see to it tomorrow.

File: ReportStats.py (hunk 22 above)
Comment (uzair-imtiaz) [reply ReportStats.py]: There's an issue with the N key. I have resolved it now.

File: ReportStats.py (hunk 23 above)
Comment (uzair-imtiaz) [reply ReportStats.py]: Fixed.

File: arg_parser.py (hunk 24 above)
Comment (uzair-imtiaz) [reply arg_parser.py]: Fixed.

File: ComputeResults.py (hunk 25 above)
Comment (uzair-imtiaz) [reply ComputeResults.py]: Related statements have been grouped with a new line separating a group from the other one.

File: ReportGenerator.py (hunk 26 above)
Comment (uzair-imtiaz) [reply ReportGenerator.py]: Alright. 
The file name has been removed.

File: .gitignore (hunk 21 above)
Comment (uzair-imtiaz) [reply .gitignore]: Some keys are too sensitive and some just won't respond.
I've removed the redundant lines.

File: .gitignore [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+# Ignore the `dist` directory if applicable
+/dist/
+
+# Ignore the `build` directory if applicable
+/build/
+
+# Ignore any sensitive or environment-specific files
+secret.txt
+config.ini
+credentials.json
```
Comment (Ali-D-Akbar) [inline .gitignore]: line ending missing

File: ReportGenerator.py [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+            Must contain at one value.
+        """
+
+        file_parser = FileParser(files[0])
+        reading_list = file_parser.parse_file()
+        weather_readings = file_parser.populate(reading_list)
+        compute = ComputeResults()
+        selected_results = compute.assign_results(weather_readings)
+        # Loop through the files of a specified month.
+        for file in files[1:]:
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py:60]: Readability can be improved.

File: ReportGenerator.py [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+        months = [
+            'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
+        ]
+        path_to_files_dir = args[0]
+        ranges = []
+        for i in range(1, len(args)):
+            ranges.append(args[i])
+
+        # Using glob.glob function to get a list of all the filenames having the given month or year as wildcard.
+        if ranges[2] is not None:
```
Comment (Ali-D-Akbar) [inline ReportGenerator.py:157]: Readability can be improved.

File: ReportStats.py [hunk 30]
```diff
@@ -0,0 +1,125 @@
...
+    # Set the extreme values and their corresponding dates
+    report_stats.set_extremes(highest, day_of_highest, lowest, day_of_lowest, humidity, most_humid_day)
+
+    # Set the average values
+    report_stats.set_averages(highest_avg_temperature, lowest_avg_temperature, avg_mean_humidity)
+"""
+
+
+class ReportStats:
+    """A class that will hold the calculations results.
```
Comment (Ali-D-Akbar) [inline ReportStats.py]: """
docstring
"""

**Review by Ali-D-Akbar — APPROVED** [review]
Remember to squash commits before merging. Let me know when you're about to merge.

**Pull request merged**